File: pulsar/java/src/main/java/com/examples/hello/pulsar/consumers/batch/BatchConsumerRunner.java
package com.examples.hello.pulsar.consumers.batch;

import com.examples.hello.pulsar.services.PulsarRunner;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.apache.pulsar.client.api.Consumer;

import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@Slf4j
@RequiredArgsConstructor
public class BatchConsumerRunner<T> extends PulsarRunner implements BatchConsumerRunnable {
  private final Consumer<T> consumer;
  private final BatchConsumer<T> sink;

  @Override
  @SneakyThrows
  protected void callOnce() {
    val messages = consumer.batchReceive();
    if (messages.size() == 0) return;
    log.debug("Received batch (size={})", messages.size());
    val msg = StreamSupport.stream(messages.spliterator(), false).collect(Collectors.toList());
    sink.accept(msg);
    consumer.acknowledge(messages);
    log.debug("Processed batch (size={})", messages.size());
  }

  @Override
  public void close() throws Exception {
    stop();
    consumer.close();
  }
}

File: pulsar/kotlin/src/main/kotlin/com/example/hello/pulsar/consumers/batch/BatchConsumerRunner.kt
package com.example.hello.pulsar.consumers.batch

import com.example.hello.pulsar.services.PulsarRunner
import org.apache.pulsar.client.api.Consumer
import org.apache.pulsar.client.api.Message
import org.slf4j.LoggerFactory

class BatchConsumerRunner<T>(
  private val consumer: Consumer<T>,
  private val sink: (List<Message<T>>) -> Unit
) : PulsarRunner(), BatchConsumerRunnable {

    override fun callOnce() {
        val messages = consumer.batchReceive()
        if (messages.size() == 0) return
        log.debug("Received batch (size={})", messages.size())
        sink.invoke(messages.toList())
        consumer.acknowledge(messages)
        log.debug("Processed batch (size={})", messages.size())
    }

    override fun close() {
        stop()
        consumer.close()
    }

  companion object {
    private val log = LoggerFactory.getLogger(BatchConsumerRunner::class.java)
  }
}
